### ISSUE TITLE
Omnisharp's mono tries to use system mono lib instead of bundled mono

### ISSUE THREAD

**Issue by benknoble** [OmniSharp/omnisharp-roslyn#1693]
I get this output when OmniSharp tries to start up in vim:

```
channel 0 open: 'dyld: Library not loaded: /Library/Frameworks/Mono.framework/Versions/6.6.0/lib/libintl.8.dylib'
channel 0 open: '  Referenced from: /Users/Knoble/.cache/omnisharp-vim/omnisharp-roslyn/bin/mono'
channel 0 open: '  Reason: image not found'
channel 0 open: '/Users/Knoble/.cache/omnisharp-vim/omnisharp-roslyn/run : ligne 28 : 54423 Abort trap: 6           "${mono_cmd}" "${omnisharp_cmd}" "$@"'
```

I can confirm that running the `run` script and the `bin/mono` fail, so this may be more for OmniSharp/omnisharp-roslyn.

I have an (old?) mono install floating around:

```
+λ ls /Library/Frameworks/Mono.framework/Versions/5.4.1/                              
Commands/   bin/        etc/        lib/        share/                                
VERSION     docs/       include/    man/        updateinfo
```

Regardless, I'm not sure why this mono is searching `/Library/Frameworks` if it was supposed to be fully bundled?

After installing an up-to-date mono with `brew`, the bundled mono still looks in `/Library/Frameworks` where the new version is not.

(From OmniSharp/omnisharp-vim#556.)

**Comment by filipw** [comment]
the minimum required Mono version is 6.6.0.
OmniSharp looks at global mono if you have it installed on your machine because it would prefer to use that msbuild instead of the bundled one.

**Comment by benknoble** [comment]
Right, but I definitely don't have mono 6.6.0 (or, I didn't...). Surely this should be easily detectable, and omnisharp could use the bundled mono?

**Comment by DwaynesWorld** [comment]
Yea, I am running in to the same issue. It happens when I am running version 1.21.10 of the C# extension in vscode (ms-vscode.csharp). If I downgrade everything seems to be fine.

**Comment by filipw** [comment]
@JoeRobich isn't there something missing in the bundle here? https://github.com/OmniSharp/omnisharp-roslyn/pull/1679/files

libintl.8.dylib?

**Comment by traegerp** [comment]
I am having the same issue as @DwaynesWorld with vs code: 

dyld: Library not loaded: /Library/Frameworks/Mono.framework/Versions/6.6.0/lib/libintl.8.dylib

**Comment by enigmaturing** [comment]
I am running into the same issue. I downgraded to 1.21.9 by the moment. Is there any update on this?

**Comment by filipw** [comment]
aa discussed elsewhere, this is caused by a mono bug.

the simplest workaround for now is to install mono 6.6.0 so that the path can be found.

**Comment by gsscoder** [comment]
@filipw,
I've exactly the same issue on a **macOS** machine. But the problem appeared in somewhat _mysterious way_.

I don't know what changed in my system and why. But obviusoly something should be happened!

Hope I can fix installing latest mono with `brew`.

**Comment by enigmaturing** [comment]
> aa discussed elsewhere, this is caused by a mono bug.
> 
> the simplest workaround for now is to install mono 6.6.0 so that the path can be found.

Thanks @filipw :)

**Comment by gsscoder** [comment]
@enigmaturing , I used __hope__ because I was not sure to have success with this.

In part I was right. I've in this moment installed **mono** with `brew`. Problem disappeared, but in all my test projects OmniSharp doesn't resolve any symbol (even `System.Object`).

No `dotnet restore` worked.

**Comment by enigmaturing** [comment]
> @enigmaturing , I used **hope** because I was not sure to have success with this.
> 
> In part I was right. I've in this moment installed **mono** with `brew`. Problem disappeared, but in all my test projects OmniSharp doesn't resolve any symbol (even `System.Object`).
> 
> No `dotnet restore` worked.
> 
> I'll file an issue if I can't solve.

@gsscoder I guess you can also downgrade to 1.21.9 by the moment, until you find a solution for that

**Comment by filipw** [comment]
> I've exactly the same issue on a **macOS** machine. But the problem appeared in somewhat _mysterious way_.
> 
> I don't know what changed in my system and why. But obviusoly something should be happened!

VS Code updates extensions automatically and this behavior is specific to C# extension 1.21.10. So it must have updated for you.

**Comment by gsscoder** [comment]
@filipw, same think I've then thought...

I didn't solved what happens with symbols, sorry I've to bother you with an new isuee...

**Comment by alexreidy** [comment]
Can't say if this is a good solution or how it might affect your system but I resolved with `brew install mono`

**Comment by ikeliec** [comment]
> > I've exactly the same issue on a **macOS** machine. But the problem appeared in somewhat _mysterious way_.
> > I don't know what changed in my system and why. But obviusoly something should be happened!
> 
> VS Code updates extensions automatically and this behavior is specific to C# extension 1.21.10. So it must have updated for you.

That was actually what happened to me. I fixed it by downgrading to 1.21.9

**Comment by TheBuzzSaw** [comment]
I had the issue on 1.21.10 on macOS. I fixed it by running `brew install mono`.

**Comment by baughmann** [comment]
What's the timeline on the fix for this? It has to do with the `1.21.10` update on my system.

**Comment by filipw** [comment]
this is resolved in 1.21.11 in prerelease https://github.com/OmniSharp/omnisharp-vscode/releases/tag/v1.21.11

**Comment by WellspringCS** [comment]
> the minimum required Mono version is 6.6.0.
OmniSharp looks at global mono if you have it installed on your machine because it would prefer to use that msbuild instead of the bundled one.
TL;DR: How should I read the above statement?

Longer version:
I'm hunting down why omnisharp isn't formatting my cs code in VSC, noticed I was getting this error:
`
[ERROR] Error: command 'dotnet.test.run' already exists
`
and found [this link](https://github.com/OmniSharp/omnisharp-vscode/issues/2952 )
That set me on a journey of discovery regarding mono -- since updating it was noted as a "good thing" for resolving the error.

And here I am.

Noob question: is it "good" (from a VSC perspective) to install mono on Ubuntu separately? Or "bad"... What I ended up doing was installing dotnet 3.1.101 to see what that would accomplish. Now my error message is gone. But I'm wondering if I should still install mono on my own.

And... yeah, my cs files still don't format. I don't know what I'm doing wrong there (or what's wrong on my machine) but that story for another day.

**Comment by filipw** [comment]
> TL;DR: How should I read the above statement?

if you have global Mono 6.6.0+, OmniSharp will use it and *should* give you better support when working with non-.NET Core projects.

> [ERROR] Error: command 'dotnet.test.run' already exists

this error is not related to the issue you are writing in 😀 normally when this happens, please grab the OmniSharp log (view -> output -> select omnisharp) and also any errors from VS Code window log (Help -> Toggle Dev Tools). Normally this error is related to conflict with other extension or unsuccessful start/restart.

> And... yeah, my cs files still don't format.

if that's the case please open a new issue here https://github.com/OmniSharp/omnisharp-vscode and fill in the details of the issue template too (repro steps, Omnisharp log, list of extensions you have installed and so on), and hopefully we can help

**Comment by WellspringCS** [comment]
>this error is not related to the issue you are writing in

Correct. Apologies, it was a (probably unneeded) explanation of how I got to this issue.

> if that's the case please open a new issue

Actually, I did. And as I was logging it up, the issue went away. I swear there are Heisenbug issues where Omnisharp is concerned. But if I say more here, I'll just clutter this thread even more. :)

**Comment by Iron-E** [comment]
> These directories exists: `/home/iron-e/Programming/ACTIVE/reef/`and `~/.omnisharp/Roslynator/src/Core/bin/Debug/netstandard2.0/Roslynator.Core.dll`.
> 
> For some reason the paths are been merged in one. So the path isn't valid.

Yeah. I notice that the first part of the path (`/home/iron-e/…`) is always equal to the `root_dir`, and I've tried this on a couple of other projects just to verify. It seems that now that the `root_dir` is being correctly resolved, it is being used as the context for _all_ paths rather than just some of them.

> > I have a roslyn extension in the folder ~/.omnisharp/… as well.
> 
> Is this [OmniSharp/omnisharp-vscode#2369](https://github.com/OmniSharp/omnisharp-vscode/issues/2369) related?

I don't think so… Roslynator has worked for me in VSCode and still works in `coc.nvim`. Seems to be something the `nvim_lsp` repo is mistakenly prepending.

If that's true, it would also explain why it can't find the `netstandard` or `core` .dll files— it'd be looking in `root_dir` instead of the root of the filesystem.

___

~~I'm still not sure if this is an issue with `nvim-lsp` or `omnisharp-rolsyn`, but I suppose a good way to tell would be if ALE has the same problem.~~

Edit: I took some time tonight to try and figure out what was happening. I figured out that the error I was seeing is a red herring, as `omnisharp-vim` throws a similar one when loading a .cs file.

I didn't manage to figure out what was going wrong, but I figured out a workaround: 

1. `dotnet clean && dotnet build` from the `root_dir` of the project (fixed a slew of errors for me, but not all of them)
2. Make sure there is __at least one__ real error in your code when loading up (and save after putting it there).
    * For example, I took out the first letter of the visibility modifier of some random method.

Then, the unwanted messages will be gone. Unfortunately, it seems that this introduces a new bug: the last set of virtual text that was present will remain there until a new error appears (note that this is completely unrelated to nvim-lua/diagnostic-nvim).
